README.md
# SinaPlot: an enhanced chart for simple and truthful representation of single observations over multiple classes

[![Build Status](https://travis-ci.org/sidiropoulos/sinaplot.svg?branch=master)](https://travis-ci.org/sidiropoulos/sinaplot)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/sinaplot)](https://cran.r-project.org/package=sinaplot)
[![CRAN_Downloads_Badge](http://cranlogs.r-pkg.org/badges/sinaplot)](https://cran.r-project.org/package=sinaplot)

The sinaplot is a data visualization chart suitable for plotting any single variable in a multiclass dataset. It is an enhanced jitter strip chart, where the width of the jitter is controlled by the density distribution of the data within each class.

## Installation

Get the released version from CRAN:

```R
install.packages("sinaplot")
```

Or the development version from github:

```R
# install.packages("devtools")
devtools::install_github("sidiropoulos/sinaplot")
```

## Webserver
In order to aid users without experience in R we also provide access to a [webserver](http://servers.binf.ku.dk:8890/sinaplot/) accepting excel sheets to produce the plots. 


NEWS
sinaplot NEWS

Changes in version 1.1.1 (2017-09-06)

- Updated CITATION.

Changes in version 1.1.0 (2017-04-10)

- Major fucntional update. Sinaplot now works with data frames, lists,
  numeric vectors and formulas, similar to boxplot and beeswarm.

- Method "density" (default) no longer bins the y-axis on based on the supplied
  "bins" parameter. Instead the inferred bins from function "density()" are
  used.

- A random seed is added as an optional argument and can be used to control the
  random sample jittering and allow for plot reproducibility.

- NA's are automatically removed. A warning is thrown if NA values are present
  in the data informing about the number of cases removed.

- "labels" parameter controls the x-axis labels. Label rotation is removed. It
  is up to the user to control the label appearance.

- sinaplot returns an invisible object regardless of "plot" parameter.
  "bin_count" column is removed from the output.

- "color" is renamed to "col" to maintain consistency with 'graphics'
  functions.

- "col" and "pch" can be used and recycled if necessary to control the color and
  the shape of the points respectively.

- sinaplot can be added on top of other plots (i.e. boxplot) by using the 'add'
  parameter.

================================================================================

Changes in version 1.0.0 (2016-10-11)

- Major graphic update. SinaPlot now uses "base" graphics instead of ggplot2.
  geom_sina() is now part of ggforce (https://github.com/thomasp85/ggforce),
  an extention of ggplot2, and can be used to produce sinaplots with ggplot2.

- Major functional update. Several parameters have been added, renamed, replaced
  or removed.
  Read below for details.

  Added:
  - "maxwidth" confines the spread of the samples along the x-axis. 0 is
  equivalent to no spread, 1 allows for maximum spread.
  - Any argument can be passed to plot(), e.g. "cex" = 0.8, "pch" = 20, etc.

  Renamed:
  - Method "neighbourhood" is renamed to "counts".
  - "neighbLimit" is renamed to "bin_limit".
  - "groupwiseScale" is renamed to "scale".

  Replaced:
  - "yFraction" is replaced with "bins". Instead of specifying the fraction of
  the data range for each bin, now you provide the number of bins.

  Removed:
  - "xSpread".
  - "labels" .
  - "main", "bw", "shape" and "size".

- If "plot == FALSE" the function returns a data frame with columns:

  x: discrete x-coordinates, split by group
  y: input values
  group: input groups
  bin_count: number of samples per bin per group
  scaled: final x-coordinates, adjusted by the sinaplot method

================================================================================

Changes in version 0.1.6 (2016-07-08)

- Fixed a bug where having samples values overlapping the outermost y-bin
  boundaries produced an error (thanks to John Hoch).

Changes in version 0.1.5 (2016-05-14)

- Fixed a bug that was plotting the first and last classes beyond the figure
  margins.

Changes in version 0.1.4 (2016-04-19)

- Fixed a bug where the data were reordered after calculating the new x-axis
  coordinates. This was creating problems when a user wanted to plot the samples
  with custom colors using sinaplot() + geom_point(col = mycol).


Changes in version 0.1.3 (2015-11-20)

- "groups" are treated as factors and they are plotted in the order that their
  levels appear (alphabetically by default) providing a similar output format with
  default ggplot2 plots.

Changes in version 0.1.2 (2015-10-02)

- NEWS file added.
- Changes in vignette compiling. knitr is added as the vignette builder in the
  DESCRIPTION FILE.
- Author details edited.
- Removed redundant "inst/temp.R" file.


inst/temp.R
mu<-2
si<-0.6
bimodal<-c(rnorm(1000,-mu,si),rnorm(1000,mu,si))
uniform<-runif(2000,-4,4)
normal<-rnorm(2000,0,3)

pdf("inst/figures/comparison.pdf", width = 10)
par(mfrow = c(1,3))

labels <- c("bimodal", "uniform", "normal")

boxplot(bimodal,uniform,normal, names = c("", "", ""), main = "Box and Whisker")
text(x = 1:length(labels), y = par("usr")[3] - 1, srt = 45, adj = 1,
     labels = c("bimodal", "uniform", "normal"), xpd = TRUE)

vioplot(bimodal,uniform,normal, names = c("", "", ""))
title("Violin")
text(x = 1:length(labels), y = par("usr")[3] - 1, srt = 45, adj = 1,
     labels = c("bimodal", "uniform", "normal"), xpd = TRUE)

sina <- sinaplot(c(bimodal, uniform, normal), c(rep("bimodal", 2000), rep("uniform", 2000),
                                        rep("normal", 2000)), bw = TRUE, scale = F, plot = F)
plot(sina[[1]], xlim = c(0,4), xaxt = "n", xlab = "", ylab = "", main = "Sina", pch = 20)
axis(1, at = 1:3, labels = FALSE)
text(x = 1:length(labels), y = par("usr")[3] - 1, srt = 45, adj = 1,
     labels = c("bimodal", "uniform", "normal"), xpd = TRUE)
dev.off()

library(beeswarm)
sinaplot(df[,2], df[,1], ytitle = "log2 expression")


p1 <- sinaplot(df[,2], df[,1], size = 1.5) + ggtitle("Sina") + theme_bw()
p1 <- p1 + geom_point(colour="black") + theme(axis.text.x = element_blank()) + ylab("log2 expression")

p2 <- ggplot(aes(V1, V2),data = df) + geom_violin(fill = "grey80") + theme_bw()
p2 <- p2 + theme(axis.text.x=element_blank()) + guides(fill=FALSE)
p2 <- p2 + ylab("") + xlab("") + ggtitle("Violin")

p3 <- ggplot(aes(V1, V2),data = df) + geom_boxplot(fill = "grey80") + theme_bw()
p3 <- p3 + theme(axis.text.x=element_blank()) + guides(fill=FALSE)
p3 <- p3 + ylab("log2 expression") + xlab("") + ggtitle("Box and Whisker")

beeswarm2 <- beeswarm(df[,2] ~ df[,1])
p4 <- ggplot(beeswarm2, aes(x, y)) + geom_point() + theme_bw()
#p4 <- ggplot(beeswarm2, aes(x, y)) + geom_point(aes(colour = df$V1)) + scale_x_discrete(limits = labels)
p4 <- p4 + theme(axis.text.x = element_blank()) + guides(colour=FALSE)
p4 <- p4 + ylab("") + xlab("") + ggtitle("Bee Swarm") + xlim(c(0.5,5.5))

p5 <- ggplot(df, aes(V1, V2)) + geom_jitter() + theme_bw()
p5 <- p5 + theme(axis.text.x=element_text()) + guides(fill=FALSE) + scale_x_discrete(limits = labels)
p5 <- p5 + ylab("log2 expression") + xlab("") + ggtitle("Jitter")

p6 <- ggplot(df, aes(V1, V2)) + geom_point() + theme_bw()
p6 <- p6 + theme(axis.text.x=element_text()) + guides(fill=FALSE) + scale_x_discrete(limits = labels)
p6 <- p6 + ylab("") + xlab("") + ggtitle("Stripchart")


pdf("inst/figures/comparison.pdf", width = 10)
png("inst/figures/comparison.png", width = 650, height = 800, res = 72)
#grid.arrange(p1, p2, p4, p3, ncol = 2)
grid.arrange(p1, p2, p3, p4, p5, p6, ncol = 2)
dev.off()
df[,1] <- factor(df[,1], levels = unique(df[,1]))

######################################
library(sinaplot)
data(blood)

a <- data$value[data$type == "CLL"]
dens <- density(a)

yFraction <- 0.04

window_size <- (max(a) - min(a)) * yFraction

yBins <- c()
for (i in 0:ceiling(1/yFraction)) {
    yBins <- c(yBins, ymin + i * window_size)
}

par(mfrow=c(1,4))
plot(rep(1, length(a)), a, xlim = c(0,2), ylim = c(4,10), ylab = "", xlab = "",
     xaxt = "n", yaxt = "n", bty="n", pch = 20)
axis(1, at = seq(0, 2, 0.5), labels = FALSE)
axis(2, at = c(4,6,8,10), labels = FALSE)


plot(rep(1, length(a)), a, xlim = c(0,2), ylim = c(4,10), ylab = "", xlab = "",
     xaxt = "n", yaxt = "n", bty="n", pch = 20)
points(dens$y + 1, dens$x, type="l")
points(-dens$y + 1, dens$x, type="l")

abline(h = yBins, lty = 2, lwd = 0.5)
axis(1, at = seq(0, 2, 0.5), labels = FALSE)
axis(2, at = c(4,6,8,10), labels = FALSE)

b <- rep(1, length(a))
p <- findInterval(a, yBins[10:11]) == 1

cur_bin <- yBins[10:11]
xMax <-mean(dens$y[findInterval(dens$x, cur_bin)==1])
xTranslate <- runif(sum(p), -xMax, xMax )
b[p] <- b[p] + xTranslate

plot(b, a, xlim = c(0,2), ylim = c(4,10), ylab = "", xlab = "",
     xaxt = "n", yaxt = "n", bty="n", pch = 20)
points(dens$y + 1, dens$x, type="l")
points(-dens$y + 1, dens$x, type="l")

abline(h = yBins, lty = 2, lwd = 0.5)
axis(1, at = seq(0, 2, 0.5), labels = FALSE)
axis(2, at = c(4,6,8,10), labels = FALSE)

c <- sinaplot(a, rep("CLL", length(a)), plot = FALSE)
plot(c[[1]]$x, c[[1]]$y, xlim = c(0,2), ylim = c(4,10), ylab = "", xlab = "",
     xaxt = "n", yaxt = "n", bty="n", pch = 20)
points(dens$y + 1, dens$x, type="l")
points(-dens$y + 1, dens$x, type="l")

abline(h = yBins, lty = 2, lwd = 0.5)
axis(1, at = seq(0, 2, 0.5), labels = FALSE)
axis(2, at = c(4,6,8,10), labels = FALSE)


R/jitterplot.R
#' @title jitterplot
#' @description
#'
#' @author Nikos Sidiropoulos
#'
#' @param x numeric vector of values to be plotted
#' @param groups vector of length x
#' @param labels optional
#' @param plot logical. Default: TRUE
#'
#' @return
#'
#'
#' @export

jitterplot <- function(x, groups, labels = NULL, plot = TRUE, ypercentage = 0.02)
{

    x <- .getXcoord(data, groups, xpercentage = 0.10, ypercentage = 0.02 )
}


.getXcoord <- function(data, groups, xpercentage, ypercentage){

    #number of groups
    ngroups <- length(unique(groups))


    #get ylim
    ymin <- min(data)
    ymax <- max(data)

    window_size <- (ymax - ymin) * ypercentage

    #
    ybreaks <- c()
    for (i in 0:ceiling(1/ypercentage)) {
        ybreaks <- c(ybreaks, ymin + i * window_size)
    }

    #
    xyArray <- c()

    #find the densiest area in the plot

    neighbours <- list()

    #set maxNeighbours
    maxNeighbours <- 0
    for (j in 1:length(levels(groups))){
        cur_xyArray <- as.data.frame(cbind(rep(j, sum(groups == levels(groups)[j])),
                                           as.numeric(data[groups == levels(groups)[j]])))
        colnames(cur_xyArray) <- c("x", "y")

        cur_neighbours <- table(findInterval(cur_xyArray$y, ybreaks))
        cur_maxNeighbours <- max(cur_neighbours)

        if (cur_maxNeighbours > maxNeighbours)
            maxNeighbours <- cur_maxNeighbours

        neighbours <- c(neighbours, list(cur_neighbours))

        xyArray <- c(xyArray, list(cur_xyArray))

    }


    for (j in 1:length(levels(groups))){

        #cur_xyArray <- as.data.frame(cbind(rep(j, sum(groups == levels(groups)[j])),
        #                     as.numeric(data[groups == levels(groups)[j]])))
        #colnames(cur_xyArray) <- c("x", "y")

        #define x-axis spacing of samples within the same neighbourhood as a
        #percentage of 1.0

        #get neighbour counts
        #neighbours <- table(findInterval(cur_xyArray$y, ybreaks))
        #maxNeighbours <- max(neighbours)

        #if the space required to spread the samples in a neighbourhood exceeds
        #1, create a scaling factor to compress the points
        if (maxNeighbours > 1/xpercentage){
            scalingFactor <- (1/xpercentage) / maxNeighbours
        }else
            scalingFactor <- 1

        #
        relScalingFactor <- max(neighbours[[j]]) / maxNeighbours

        for (i in names(neighbours[[j]])){

            if (neighbours[[j]][i] > 1){
                cur_break <- ybreaks[ as.integer(i) : (as.integer(i) + 1)]

                break_levels <- levels(cut(1:100*xpercentage*neighbours[[j]][i],
                                           neighbours[[j]][i]))
                breaks <- cbind(lower = as.numeric( sub("\\((.+),.*", "\\1",
                                                        break_levels)),
                                upper = as.numeric( sub("[^,]*,([^]]*)\\]",
                                                        "\\1", break_levels)))
                xTranslate <- apply(breaks, 1, mean) - mean(apply(breaks, 1, mean))
                #neighbours[i]

                points <- findInterval(xyArray[[j]]$y, cur_break) == 1
                xyArray[[j]]$x[points] <- xyArray[[j]]$x[points] +
                    (xTranslate * scalingFactor * relScalingFactor)/ 100
            }
        }


    }

    do.call(rbind, xyArray)

}



R/sinaplot.R
#' @title sinaplot
#'
#' @description The SinaPlot is a data visualization chart suitable for plotting
#' any single variable in a multiclass dataset. It is an enhanced jitter strip
#' chart, where the width of the jitter is controlled by the density
#' distribution of the data within each class.
#'
#' @param x numeric vector or a data frame or a list of numeric vectors to be
#' plotted.
#'
#' @param ... arguments to be passed to \code{\link[graphics]{plot}}.
#'
#' @details There are two available ways to define the x-axis borders for the
#' samples to spread within:
#' \itemize{
#'  \item{\code{method = "density"}
#'
#'   A density kernel is estimated along the y-axis for every sample group. The
#'   borders are then defined by the density curve. Tuning parameter
#'   \code{adjust} can be used to control the density bandwidth in the same way
#'   it is used in \code{\link[stats]{density}}. }
#'
#'  \item{\code{method = "counts"}:
#'
#'  The borders are defined by the number of samples that occupy the same
#'  bin and the parameter \code{maxwidth} in the following fashion:
#'
#'  \code{xBorder = nsamples * maxwidth}
#'
#'   }
#' }
#'
#' @return
#'
#' \item{x}{discrete x-coordinates, split by group}
#' \item{y}{input values}
#' \item{group}{input groups}
#' \item{scaled}{final x-coordinates, adjusted by sinaplot}
#'
#' @examples
#'
#' ## sinaplot on a formula:
#'
#' data("blood", package = "sinaplot")
#' boxplot(Gene ~ Class, data = blood)
#' sinaplot(Gene ~ Class, data = blood, pch = 20, add = TRUE)
#'
#' ## sinaplot on a data.frame:
#'
#' df <- data.frame(Uni05 = (1:100)/21, Norm = rnorm(100),
#'                   `5T` = rt(100, df = 5), Gam2 = rgamma(100, shape = 2))
#' boxplot(df)
#' sinaplot(df, add = TRUE, pch = 20)
#'
#' ## sinaplot on a list:
#'
#' bimodal <- c(rnorm(300, -2, 0.6), rnorm(300, 2, 0.6))
#' uniform <- runif(500, -4, 4)
#' normal <- rnorm(800,0,3)
#'
#' distributions <- list(uniform = uniform, bimodal = bimodal, normal = normal)
#' boxplot(distributions, col = 2:4)
#' sinaplot(distributions, add = TRUE, pch = 20)
#'
#' ## sinaplot on a vector:
#'
#' x <- c(rnorm(200, 4, 1), rnorm(200, 5, 2), rnorm(400, 6, 1.5))
#' groups <- c(rep("Cond1", 200), rep("Cond2", 200), rep("Cond3", 400))
#'
#' sinaplot(x, groups)
#'
#' par(mfrow = c(2, 2))
#'
#' sinaplot(x, groups, pch = 20, col = 2:4)
#' sinaplot(x, groups, scale = FALSE, pch = 20, col = 2:4)
#' sinaplot(x, groups, scale = FALSE, adjust = 1/6, pch = 20, col = 2:4)
#' sinaplot(x, groups, scale = FALSE, adjust = 3, pch = 20, col = 2:4)
#'
#' #blood
#'
#' par(mfrow = c(1,1))
#' sinaplot(blood$Gene, blood$Class)
#'
#' old.mar <- par()$mar
#' par(mar = c(9,4,4,2) + 0.1)
#' groups <- levels(blood$Class)
#'
#' sinaplot(blood$Gene, blood$Class, pch = 20, xaxt = "n", col = rainbow(18))
#' axis(1, at = 1:length(groups), labels = FALSE)
#' text(1:length(groups), y = par()$usr[3] - 0.1 * (par()$usr[4] - par()$usr[3]),
#'      xpd = TRUE, srt = 45, adj = 1, labels = groups)
#' par(mar = old.mar)
#'
#' @rdname sinaplot
#' @export

sinaplot <- function (x, ...)
    UseMethod("sinaplot")

#' @return \code{NULL}
#'
#' @param groups optional vector of \code{length(x)}.
#'
#' @param method choose the method to spread the samples within the same
#' bin along the x-axis. Available methods: "density" and "counts".
#' See \code{Details}.
#'
#' @param scale a logical that indicates whether the width of each group should
#' be scaled relative to the group with the highest density.
#' Default: \code{TRUE}.
#'
#' @param adjust adjusts the bandwidth of the density kernel when
#' \code{method == "density"} (see \code{\link[stats]{density}}).
#'
#' @param bins number of bins to divide the y-axis into when
#' \code{method == "counts"}. Default: 50.
#'
#' @param bin_limit if the samples within the same y-axis bin are more
#' than \code{bin_limit}, the samples's X coordinates will be adjusted.
#'
#' @param maxwidth control the maximum width the points can spread into. Values
#' between 0 and 1.
#'
#' @param seed a single value that controls the random sample jittering. Set to
#' an integer to enable plot reproducibility. Default NULL.
#'
#' @param plot logical. When \code{TRUE} the sinaplot is produced, otherwise the
#' function returns the new sample coordinates. Default: \code{TRUE}.
#'
#' @param add logical. If true add boxplot to current plot.
#'
#' @param log logical. If true it uses a logarihmic scale on the y-axis.
#'
#' @param labels labels for each group. Recycled if necessary. By default,
#' these are inferred from the data.
#'
#' @param xlab,ylab axis labels.
#'
#' @param pch,col plotting characters and colors, specified by group.
#' Recycled if necessary.
#'
#' @rdname sinaplot
#' @method sinaplot default
#' @importFrom plyr ddply mutate
#' @importFrom graphics points axis par text box
#' @importFrom stats complete.cases na.omit
#' @export
sinaplot.default <- function(x,
                     groups = NULL,
                     method = c("density", "counts"),
                     scale = TRUE,
                     adjust = 0.75,
                     bins = 50,
                     bin_limit = 1,
                     maxwidth = 1,
                     seed = NULL,

                     #Plot parameters
                     plot = TRUE,
                     add = FALSE,
                     log = FALSE,
                     labels = NULL,
                     xlab = "",
                     ylab = "",
                     col = NULL,
                     pch = NULL,
                     ...
                     ) {

    ###Check input arguments
    if (!is.null(groups)) {
        if (!is.numeric(x))
            stop("x must be a numeric vector if 'groups' are provided")

        if (length(x) != length(groups))
            stop("x and groups must be of the same length.")
    }

    if (bin_limit < 1 | !is.numeric(bin_limit)){
        warning("Invalid bin_limit value. bin_limit was set to 1.")
        bin_limit <- 1
    }

    if (!is.logical(log)) {
        warning("log must be TRUE/FALSE. log is set to FALSE")
        log <- FALSE
    }

    if (bins <= 0) {
        warning("bins must be > 0. Set to 50.")
        bins <- 50
    }

    if (maxwidth < 0 | maxwidth > 1){
        warning("maxwidth must be between 0 and 1. maxwidth set to 1.")
        maxwidth <- 1
    }

    if (is.null(names(x))) {
        x.names <- FALSE
    } else {
        x.names <- TRUE
    }

    if (is.null(groups)) {
        if (is.numeric(x)) {

            groups <- 1

        } else {

            if (x.names) {
                groups <- rep(names(x), sapply(x, length))
            } else {
                groups <- rep(1:length(x), sapply(x, length))
            }

            x <- unlist(x, use.names = FALSE)

        }
    } else {
        x.names <- TRUE
    }

    bin <- NULL

    method <- match.arg(method)
    ###end

    #remove redundant labels
    if (!is.factor(groups))
        groups <- factor(groups, levels = unique(groups))

    data <- data.frame(x = as.numeric(groups), y = x, group = groups,
                       x_translation = 0)

    #remove NA's
    count_na <- sum(!complete.cases(data))
    if (count_na > 0) {

        warning(paste("Removed", count_na,
                      "rows containing non-finite values (sinaplot)"),
                sep = " ")

        data <- na.omit(data)
    }

    data <- ddply(data, "group", .compute, method, maxwidth, adjust, bin_limit,
                  bins, seed)

    if (scale) {
        group_max  <- ddply(data, "group", mutate,
                            group_max = max(table(bin)))$group_max
        group_scaling_factor <- group_max / max(group_max)
    } else {
        group_scaling_factor <- 1
    }

    data$scaled <- data$x + data$x_translation * group_scaling_factor

    #drop columns
    data$x_translation <- NULL
    data$bin <- NULL

    n.groups <- length(levels(groups))

    if(missing(labels) || is.null(labels)) {
        if(!x.names) {
            if(n.groups == 1) {
                labels <- NA
            } else {
                labels <- 1:n.groups
            }
        } else {
            labels <- levels(groups)
        }
    } else {
        labels <- rep(labels, length.out = n.groups)
    }

    if (is.null(col))
        col <- par("col")
    else
        col <- rep(rep(col, length.out = n.groups), times = table(groups))

    if (is.null(pch))
        pch <- par("pch")
    else
        pch <- rep(rep(pch, length.out = n.groups), times = table(groups))

    data$col <- col
    data$pch <- pch

    ylim <- range(data$y[is.finite(data$y)])

    if (log) {
        if (any(ylim < 0)) {
            warning(paste("Logarithmic axis must have positive limits.",
                          "Log set to FALSE.", sep = " "))
            ylog <- ""
        } else
            ylog <- "y"
    } else
        ylog <- ""

    if (plot){

        if (!add) {

            xlim <- c(0.5, length(levels(data$group)) + 0.5)

            plot(xlim, ylim, type = 'n', axes = FALSE, log = ylog,
                     xlab = xlab, ylab = ylab, ...)
        }

        points(data$scaled, data$y, col = col, pch = pch, ...)
        axis(1, at = 1:n.groups, labels = labels, ...)
        axis(2, ...)
        box(...)
    }

    invisible(data)
}

#' @return \code{NULL}
#' @rdname sinaplot
#'
#' @param formula a formula, such as y ~ grp, where y is a numeric vector of
#' data values to be split into groups according to the grouping variable grp
#' (usually a factor).
#' @param data a data.frame (or list) from which the variables in formula should
#' be taken.
#' @param subset an optional vector specifying a subset of observations to be
#' used for plotting.
#' @param na.action a function which indicates what should happen when the data
#' contain NAs. The default is to ignore missing values in either the response
#' or the group.
#' @method sinaplot formula
#' @export
sinaplot.formula <-
    function(formula, data = NULL, ..., subset, na.action = NULL,
             xlab, ylab)
    {
        if(missing(formula) || (length(formula) != 3L))
            stop("'formula' missing or incorrect")
        m <- match.call(expand.dots = FALSE)
        if(is.matrix(eval(m$data, parent.frame())))
            m$data <- as.data.frame(data)
        m$... <- NULL
        m$xlab <- NULL
        m$ylab <- NULL
        m$na.action <- na.action # force use of default for this method
        ## need stats:: for non-standard evaluation
        m[[1L]] <- quote(stats::model.frame)
        mf <- eval(m, parent.frame())
        response <- attr(attr(mf, "terms"), "response")
        if (missing(xlab))
            xlab <- as.character(formula)[3]
        if (missing(ylab))
            ylab <- names(mf)[response]
        sinaplot(split(mf[[response]], mf[-response]), xlab = xlab, ylab = ylab,
                 ...)
}

.compute <- function(data, method, maxwidth, adjust, bin_limit, n_bins, seed) {
    #initialize x_translation and bin_counts to 0

    #if group has less than 2 points return as is
    if (nrow(data) < 2) {
        data$bin <- 1
        data
    }

    #per bin sample density
    if (method == "density") {
        density <- stats::density(data$y, adjust = adjust, na.rm = TRUE)

        #confine the samples in a (-maxwidth/2, -maxwidth/2) area around the
        #group's center
        intra_scaling_factor <- 0.5 * maxwidth / max(density$y)

        # assign data points to density bins
        data$bin <- findInterval(data$y, density$x)

        # jitter points based on the density
        if (!is.null(seed))
            set.seed(seed)
        x_translation <- sapply(density$y[data$bin], jitter, x = 0, factor = 1)

        #scale and store new x coordinates
        data$x_translation <- x_translation * intra_scaling_factor

    } else {

        bins <- .bin_y(range(data$y), n_bins)
        #per bin sample count
        data$bin <- findInterval(data$y, bins)

        bin_counts <- table(data$bin)

        #allow up to 50 samples in a bin without scaling
        intra_scaling_factor <- 50 * maxwidth / max(bin_counts)

        for (i in names(bin_counts)) {
            #examine bins with more than 'bin_limit' samples
            if (bin_counts[i] > bin_limit){
                cur_bin <- bins[ as.integer(i) : (as.integer(i) + 1)]

                #find samples in the current bin and translate their X coord.
                points <- findInterval(data$y, cur_bin) == 1

                #compute the border margin for the current bin.
                xmax <- bin_counts[i] / 100

                #assign the samples uniformely within the specified range
                if (!is.null(seed))
                    set.seed(seed)
                x_translation <- stats::runif(bin_counts[i], - xmax, xmax)

                #scale and store new x coordinates
                data$x_translation[points] <-
                    x_translation * intra_scaling_factor
            }
        }
    }

    data
}

.bin_y <- function(data, bins) {
    #get y value range
    ymin <- min(data)
    ymax <- max(data)

    #window width
    window_size <- (ymax - ymin) / (bins + 1e-8)

    ybins <- c()
    for (i in 0:ceiling(bins)) {
        ybins <- c(ybins, ymin + i * window_size)
    }

    ybins
}
